vllm/attention/backends/dual_chunk_flash_attn.py

```diff
@@ -0,0 +1,877 @@
+            ke = cu_seqlens_k_cpu[i:i + 2][-1]
+
+            current_q = q[qs:qe]
+            current_q_succ = q_succ[qs:qe]
+            current_q_inter = q_inter[qs:qe]
+            if block_table is None:
+                current_k = k[ks:ke]
+                current_v = v[ks:ke]
+                current_block_table = None
+                current_prefill_original_seq_lens_tensor = (
+                    prefill_original_seq_lens_tensor[i:i + 1])
+            else:
+                current_block_table = block_table[i:i + 1]
+                current_prefill_original_seq_lens_tensor = (
+                    prefill_original_seq_lens_tensor[i:i + 1])
+                current_k = k
+                current_v = v
+
+            if current_q.shape[0] == 0:
+                continue
+            if current_k.shape[0] == 0:
+                all_outputs.append(
+                    torch.zeros(
+                        (current_q.shape[0], current_q.shape[1], v.shape[2]),
+                        device=q.device,
+                        dtype=q.dtype,
+                    ))
+                continue
+
+            current_output = self._bruteforce_dynamic_chunk_flash_attn_func(
+                current_q,
+                current_q_succ,
+                current_q_inter,
+                current_k,
+                current_v,
+                current_block_table,
+                softmax_scale,
+                chunk_size,
+                local_size,
+                original_max_position_embeddings,
+                current_prefill_original_seq_lens_tensor,
+                ke - ks,
+            )
+            all_outputs.append(current_output)
+
+        return torch.cat(all_outputs, dim=0)
+
+    def _bruteforce_dynamic_chunk_pageattention_forward_decode(
+        self,
+        query: torch.Tensor,
+        query_succ: torch.Tensor,
+        query_inter: torch.Tensor,
+        key_cache: torch.Tensor,
+        value_cache: torch.Tensor,
+        block_table: torch.Tensor,
+        cache_seqlens: torch.Tensor,
+        softmax_scale: float,
+        causal: bool,
+        alibi_slopes: Optional[torch.Tensor],
+        chunk_size: int,
+        local_size: int,
+        original_max_position_embeddings: int,
+    ):
+        assert causal
+        batch_size = block_table.shape[0]
+        block_size = value_cache.shape[1]
+        chunk_len = chunk_size - local_size
+        if chunk_len % block_size != 0:
+            raise ValueError("chunk_len must be divisible by block_size.")
+        chunk_num_curr = (cache_seqlens - 1) // chunk_len
+
+        if original_max_position_embeddings > 0:
+            mscale = (
+                0.1 *
+                torch.log(cache_seqlens / original_max_position_embeddings) +
+                1.0).clip(min=1)
+            query = (query * mscale.view(-1, 1, 1, 1)).to(
+                query.dtype
+            )  # possible for numerical issue, need to fused in the kernel
```

Comment:
hi @hzhwcmhf, could you explain this numerical issue and why kernel fusion can solve it?

Reply:
@rejoicesyc We want to compute `(query @ key) * softmax_scale * mscale` as the attention weight, where `query` is represented in `float16` or `bfloat16`, and `mscale` is represented in `float32`. Better precision can be achieved if the scaling is computed in `float32`, rather than multiplying `query` by `mscale` before the attention operation. The flash attention kernel provides an argument `softmax_scale`, which is close to our requirement. However, `softmax_scale` is a constant for the entire query, whereas `mscale` is a vector that specifies different `softmax_scale` values for each query vector at different positions.